rfcs/006-package-management.md

```diff
@@ -0,0 +1,560 @@
+  because a new version became available
+
+I don't have much experience using this method, but it seems like one
+disadvantage is that it puts an extra burden on package maintainers to keep
+backwards compatibility even across "major" version bumps.
+
+[bzlmod] also uses some version of minimal version selection, but they
+do apparently allow for backwards-compatibility breakage. They have
+an extra piece of metadata called "compatibility level" (not encoded in
+the version number). They don't allow a dependency graph to contain different
+compatibility levels of the same module.
+
+[minimal version selection]: https://research.swtch.com/vgo-mvs
+[bzlmod]: https://docs.bazel.build/versions/5.0.0/bzlmod.html
+
+### Some interpolation between MVS and semver
+
+There are some middle grounds between "minimal version selection" and
+cargo-style resolution. We could follow semver and allow multiple incompatible
+versions of packages, while also doing one or both of the following
+
+- preferring to resolve minimal versions instead of maximal versions. This promotes
+  stability in version resolution, at the cost of requiring intervention to bring
+  in bugfixes in dependencies.
+- only allowing `^` version specifications (along with `=` overrides at the top level).
+  This amounts to being opinionated that "the ecosystem is using semver, and your
+  version requirements should comply with semver." It's basically a semver analogue
+  of go's policy of only allowing lower bounds (plus pinning at the top level).
+
+### Question: how to handle pre-1.0 minor versions?
+
+Officially, semver says that pre-1.0 versions are mutually incompatible. If we
+follow this, pre-1.0 versions would never get binned together. `cargo` modifies
+the semver rules, allowing `0.x.y` versions to be binned together if they share
+the same `x`. Should we do the same?
+
+## Yanking packages
+
+What if someone messes up and publishes a package that's incompatible with
+an earlier version? In this case, we allow packages in the registry to be
+"yanked," meaning that they are left there but marked as broken. The resolver
+will ignore yanked packages unless they are already in the lock-file.
+This means that if you're happily using a package version that got yanked
+(maybe you're not using the broken part of the package) then you won't
+get broken by the yank.
+
+Package yanking will be done with the same process as package publishing
+(i.e., probably PRs to the index). See the next section.
+
+### Alternative: allow yanked packages if resolution would otherwise fail
+
+A slightly more permissive alternative would be to ignore yanked packages that
+aren't in the lock-file, but if resolution fails then go back and try allowing
+them. This would allow resolution to succeed, for example, if your manifest
+asks for the latest version of a package *and* that package was yanked *and*
+you don't have a lock-file. But this seems like maybe adding too many corner-cases.
```

Comment:
I wonder how useful this is. That is, I don't expect it would happen a lot that there's a unique version that can be picked and it's a yanked version. Often either the yanked version is immediately followed by a backward-compatible update - minor or patch -, or there it wasn't a `X.0.0` version and there's other versions of the same major version. I think a yanked version would be the only choice if people started to depend on this specific version too fast (otherwise a previous version of the same bin could still be picked) AND the correction is a breaking change (otherwise the more up-to-date correction could still be picked).